snips_nlu/intent_parser/lookup_intent_parser.py

```diff
@@ -0,0 +1,403 @@
+from snips_nlu.exceptions import IntentNotFoundError, LoadingError
+from snips_nlu.intent_parser.intent_parser import IntentParser
+from snips_nlu.pipeline.configs import LookupIntentParserConfig
+from snips_nlu.preprocessing import normalize_token, tokenize, tokenize_light
+from snips_nlu.resources import get_stop_words
+from snips_nlu.result import (
+    empty_result,
+    intent_classification_result,
+    parsing_result,
+    unresolved_slot,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@IntentParser.register("lookup_intent_parser")
+class LookupIntentParser(IntentParser):
+    """Intent parser using pattern matching in a deterministic manner
+
+    This intent parser is very strict by nature, and tends to have a very good
+    precision but a low recall. For this reason, it is interesting to use it
+    first before potentially falling back to another parser.
+    """
+
+    config_type = LookupIntentParserConfig
+
+    def __init__(self, config=None, **shared):
+        """The deterministic intent parser can be configured by passing a
+        :class:`.DeterministicIntentParserConfig`"""
+        super(LookupIntentParser, self).__init__(config, **shared)
+        self._language = None
+        self.stop_words = None
+        self.map = None
+        self.intents_names = []
+        self.slots_names = []
+        self._intents_mapping = {}
+        self._slots_mapping = {}
+
+    @property
+    def language(self):
+        return self._language
+
+    @language.setter
+    def language(self, value):
+        self._language = value
+        if value is None:
+            self.stop_words = None
+        else:
+            if self.config.ignore_stop_words:
+                self.stop_words = get_stop_words(self.resources)
+            else:
+                self.stop_words = set()
+
+    @property
+    def fitted(self):
+        """Whether or not the intent parser has already been trained"""
+        return self.map is not None
+
+    @log_elapsed_time(
+        logger, logging.INFO,
+        "Fitted dict deterministic parser in {elapsed_time}"
+    )
+    def fit(self, dataset, force_retrain=True):
+        """Fits the intent parser with a valid Snips dataset"""
+        logger.info("Fitting dict deterministic parser...")
+        dataset = validate_and_format_dataset(dataset)
+        self.load_resources_if_needed(dataset[LANGUAGE])
+        self.fit_builtin_entity_parser_if_needed(dataset)
+        self.fit_custom_entity_parser_if_needed(dataset)
+        self.language = dataset[LANGUAGE]
+        self.map = dict()
+        entity_placeholders = _get_entity_placeholders(dataset, self.language)
+
+        for (key, val) in self.generate_io_mapping(
+                dataset[INTENTS], entity_placeholders
+        ):
+            # handle key collisions -*- remove ambiguous entries -*-
+            if key in self.map and self.map[key] != val:
+                self.map.pop(key)
+            else:
+                self.map[key] = val
```

Comment:
This does not handle collisions properly for two reasons:
- collisions between utterances of the same intent are fine
- the logic is wrong in certain cases, for instance when you have 3 collisions from 3 different intents

You should look at what has been done [here](https://github.com/snipsco/snips-nlu/blob/develop/snips_nlu/intent_parser/deterministic_intent_parser.py#L140-L158).